### PR TITLE
Fixed passive scalar + AMR bug

### PR DESCRIPTION
**Pull request by c-white**
I noticed something wrong with the passive scalars near refinement boundaries in an AMR calculation. After digging through the code, I realized the scalars were not perfectly paralleling the hydro quantities with regard to primitive vs. conserved coarse buffers. In particular, calls to `HydroBoundaryVariable::SwapHydroQuantity` were followed by the appropriate setting of `var_cc` in the scalars' `CellCenteredBoundaryVariable`, but the subsequent setting of `coarse_buf` done by `HydroBoundaryVariable::SelectCoarseBuffer` was omitted. This occurs 7 times in `mesh.cpp`, 4 times in `time_integrator.cpp`, and 2 times in `sts_task_list.cpp`.

This pull request adds in the missing logic in all 13 cases. I don't know if this is strictly necessary in all cases, but it doesn't seem like this can harm anything. I am particularly uncertain about STS, since I never use it.

This problem is hard to find in existing test problems. Partly, this may be because the effects of bad ghost cells may be hidden by flux corrections and time evolution. It may also not show up in problems where the conserved density is 1, since the primitive and conserved scalars are the same in that case. It may well affect SMR calculations, but I can't be certain I've seen that occur.

Below is an example problem before and after this update. The hydro variables always look good, but the passive scalars show clear grid artifacts in the unfixed version.

amr_scalar_comparison

- [x] Code follows style guide.
- [x] Documentation updated (N/A).
- [ ] Test coverage: I could not find a simple test that illustrated the problem (I only saw it with a 3D AMR multigrid calculation with at least 512 8^3 blocks).
- [ ] Optional: someone can comment on whether all the changes here are necessary.
- [ ] Optional: one could achieve the same effect by making a new `ScalarBoundaryVariable` derived from `CellCenteredBoundaryVariable` with analogous `SwapScalarQuantity` and `SelectCoarseBuffer` functions, more closely paralleling hydro.

### CODE REVIEW

**Review by pdmullen — COMMENTED** [review]
It looks like @c-white may have just brought #365 to its grave.  

As far as STS goes, it should just invoke the exact same infrastructure as `time_integrator.cpp`, which looks like what you have done here.  Nice find and happy Memorial Day weekend 😸

**Review by tomidakn — APPROVED** [review]

**Comment by felker** [comment]
This is an amazing find @c-white . The changes make sense to me (worst case, some of the swapping of coarse buffer variable references is redundant, but shouldnt be a problem). 

I am more surprised how they were completely missing from my original implementation. And why my design decisions did not end up with making a subfolder `src/bvals/cc/scalar` and associated `CellCenteredBoundaryVariable` derived class, like you suggest. I suppose I envisioned passive scalars as truly the generic concrete example of a `CellCenteredBoundaryVariable`, and by the time I got through the mess of writing https://github.com/PrincetonUniversity/athena/blob/master/src/bvals/bvals_refine.cpp, I didnt revisit that assumption. Also surprised how the failures were so small, although that could be attributed to only introducing problems at boundaries with **coarser** MeshBlocks (and the difference would be large only when density is far from 1.0?).

Certainly there are several unique characteristics of Hydro variables as boundary variables (mostly shearing box, orbital advection related), but this comment is definitely untrue and should be changed:
https://github.com/PrincetonUniversity/athena/blob/afcd297a9b2b8e91ae06c0637911d49f8eba081d/src/bvals/cc/hydro/bvals_hydro.hpp#L59-L60

In other words, much of https://github.com/PrincetonUniversity/athena/blob/master/src/bvals/cc/hydro/bvals_hydro.cpp and https://github.com/PrincetonUniversity/athena/blob/afcd297a9b2b8e91ae06c0637911d49f8eba081d/src/mesh/mesh_refinement.cpp#L1071 should be modified to have generic implementations that work for hydro variables and passive scalars, and any cell-centered variable that has some kind of notion of conserved and primitive variable duality that should be handled nearly the same way by AMR.

**Review by felker — APPROVED** [review]
